### PR TITLE
add the tensor_observer to record the runtime tensor for quantization …

### PR DESCRIPTION
**Pull request by lly-zero-one**
…accuracy analsyis

### CODE REVIEW

File: torch/quantization/observer.py [hunk 1]
```diff
@@ -139,6 +139,23 @@ def calculate_qparams(self):
     def extra_repr(self):
         return 'min_val={}, max_val={}'.format(self.min_val, self.max_val)
 
+
+class DebugObserver(ObserverBase):
```
Comment (hx89) [inline torch/quantization/observer.py]: Maybe rename it to TensorObserver to be more general since it's mainly observing tensor values, which could be used for other purposes than debugging :-)

**Comment by llyfacebook** [comment]
@hx89  According to our discussion, I will leave the work of wrapping top module and calling into existing add_observer to you.

File: torch/quantization/observer.py [hunk 2]
```diff
@@ -147,6 +164,9 @@ def default_observer(**kwargs):
     kwargs.setdefault("reduce_range", True)
     return observer(MinMaxObserver, **kwargs)
 
+def default_tensor_observer(**kwargs):
```
Comment (jerryzh168) [inline torch/quantization/observer.py]: maybe default_debug_observer?

File: torch/quantization/QConfig.py [hunk 3]
```diff
@@ -9,6 +9,9 @@
 default_qconfig = QConfig(default_weight_observer(),
                           default_observer())
 
+default_debug_qconfig = QConfig(default_weight_observer(),
```
Comment (jerryzh168) [inline torch/quantization/QConfig.py]: also I'm changing the order of activation and weight in qconfig, please use keyword arguments instead: https://github.com/pytorch/pytorch/pull/25950

File: test/test_quantization.py [hunk 4]
```diff
@@ -765,5 +765,25 @@ def test_observer_scriptable(self):
...
+                 ' well with UBSAN at the moment, so we skip the test if'
+                 ' we are in a UBSAN environment.')
+class QuantizationDebugTest(QuantizationTestCase):
+    def test_tensor_observer(self):
+        model = SingleLayerLinearModel()
+        model.qconfig = default_debug_qconfig
+        prepare(model)
+        # run the evaluation and dump all tensors
+        test_only_eval_fn(model, self.calib_data)
+        test_only_eval_fn(model, self.calib_data)
```
Comment (jerryzh168) [inline test/test_quantization.py]: why do you need to run this twice?

Comment (jerryzh168) [reply test/test_quantization.py]: oh I see

**Review by hx89 — CHANGES_REQUESTED** [review]
Could you add a test to test scriptability of the observer? May use the test_tensor_observer_scriptable() in #25864.

File: torch/quantization/QConfig.py [hunk 5]
```diff
@@ -9,6 +9,9 @@
 default_qconfig = QConfig(default_weight_observer(),
                           default_observer())
 
+default_debug_qconfig = QConfig(weight=default_weight_observer(),
+                          activation=default_debug_observer())
```
Comment (jerryzh168) [inline torch/quantization/QConfig.py]: nit: indentation

File: torch/quantization/quantize.py [hunk 6]
```diff
@@ -333,3 +333,23 @@ def swap_module(mod, mapping):
         if type(mod) in mapping:
             new_mod = mapping[type(mod)].from_float(mod)
     return new_mod
+
+def dump_tensor(mod, prefix, target_dict):
```
Comment (jerryzh168) [inline torch/quantization/quantize.py]: I don't think we need to expose prefix to user, we can default to '' and user can just call it with dump_tensor(mod, target_dict)

File: torch/quantization/quantize.py [hunk 7]
```diff
@@ -333,3 +333,23 @@ def swap_module(mod, mapping):
...
+    Args:
+        mod: the top module we want to save all tensors
+        prefix: the prefix for the current module
+        target_dict: the dictionary used to save the tensors
+    """
+    weight_unpack = getattr(mod, "weight", None)
+    if weight_unpack is not None and callable(weight_unpack):
+        target_dict[prefix + '.' + 'weight'] = mod.weight()
+    elif hasattr(mod, 'weight'):
+        target_dict[prefix + '.' + 'weight'] = mod.weight
```
[the remark this answers is not in the document]
Comment (llyfacebook) [reply torch/quantization/quantize.py]: For qconv/qlinear, it is callable, since we need to unpack it.

File: test/test_quantization.py [hunk 8]
```diff
@@ -765,5 +765,41 @@ def test_observer_scriptable(self):
...
+        tensor_dict = {}
+        dump_tensor(model, tensor_dict, "$")
```
Comment (jerryzh168) [inline test/test_quantization.py]: I think we can remove "$" in the test now, that doesn't feel like adding anything to the test.

**Review by jerryzh168 — APPROVED** [review]
LGTM, please address the comment in the test

**Review by hx89 — APPROVED** [review]
LGTM!

File: test/test_quantization.py (hunk 8 above)
Comment (llyfacebook) [reply test/test_quantization.py]: All right, actually, i just don't want the name starting with ".".

File: torch/quantization/quantize.py [hunk 9]
```diff
@@ -333,3 +333,26 @@ def swap_module(mod, mapping):
...
+    This is mainly used for quantization accuracy debug
+    Args:
+        mod: the top module we want to save all tensors
+        prefix: the prefix for the current module
+        target_dict: the dictionary used to save the tensors
+    """
+    def get_prefix(prefix):
+        return prefix if prefix == "" else prefix + '.'
+
+    weight_unpack = getattr(mod, "weight", None)
```
Comment (dzhulgakov) [inline torch/quantization/quantize.py]: why do we need to handle the weights part here and hard-code the names of parameters? module.state_dict() should give you what you need for this debug

File: torch/quantization/quantize.py [hunk 10]
```diff
@@ -333,3 +333,26 @@ def swap_module(mod, mapping):
...
+    if weight_unpack is not None and callable(weight_unpack):
+        target_dict[get_prefix(prefix) + 'weight'] = mod.weight()
+    elif hasattr(mod, 'weight'):
+        target_dict[get_prefix(prefix) + 'weight'] = mod.weight
+
+    if hasattr(mod, 'observer'):
+        target_dict[get_prefix(prefix) + 'activation'] = mod.observer.get_tensor_value()
```
Comment (dzhulgakov) [inline torch/quantization/quantize.py]: this blindly assumes that all observers are debug observers. What if it's not the case?

File: torch/quantization/quantize.py [hunk 11]
```diff
@@ -333,3 +333,26 @@ def swap_module(mod, mapping):
...
+    for name, child in mod.named_children():
+        module_prefix = get_prefix(prefix) + name if prefix else name
+        dump_tensor(child, target_dict, module_prefix)
```
Comment (dzhulgakov) [inline torch/quantization/quantize.py]: you could also use module.apply() I think

File: torch/quantization/quantize.py [hunk 12]
```diff
@@ -333,3 +333,26 @@ def swap_module(mod, mapping):
         if type(mod) in mapping:
             new_mod = mapping[type(mod)].from_float(mod)
     return new_mod
+
+def dump_tensor(mod, target_dict, prefix=""):
```
Comment (dzhulgakov) [inline torch/quantization/quantize.py]: I'm not sure this function is helpful beyond state_dict and manual manipulation.

Also it's name is very non-descriptive - what does 'dump_tensor' mean if I just see it?

Here's a better idea - we can implement a function `all_observers_dict(module)` that acts like `module.state_dict()` but returns you the named dictionary of all observers. Then you can just call that function and write a one-liner that iterates over observers and pulls out tensor values

File: torch/quantization/observer.py [hunk 13]
```diff
@@ -139,6 +139,32 @@ def calculate_qparams(self):
     def extra_repr(self):
         return 'min_val={}, max_val={}'.format(self.min_val, self.max_val)
 
+
+class TensorObserver(ObserverBase):
```
Comment (dzhulgakov) [inline torch/quantization/observer.py]: I'd call it RecordingObserver or something like it

**Review by dzhulgakov — COMMENTED** [review]
Sorry for coming in late - but I feel this API is a bit hacky. Please see inline on how to structure it better
